### DISCUSSION TITLE
how to set the userdata to a CURL*, i malloc a buffer, and add to the queue, then i send the data, i need free the buff, before curl_easy_cleanup

### DISCUSSION THREAD

**Discussion by xinsuinizhuan** [curl/curl#9567] [Q&A]
how to set the userdata to a CURL*, i malloc a buffer, and add to the queue, then i send the data, i need free the buff, before curl_easy_cleanup

**Comment by bagder** [comment]
I already told you the ways in https://github.com/curl/curl/discussions/9557#discussioncomment-3698041 for when curl releases that memory.

But frankly: I don't think you do. You cannot remove it before the transfer is complete, and once it has completed you might as well kill the handle.

**Reply by bagder** [reply]
Or did you mean some other data?

**Reply by xinsuinizhuan** [reply]
Yes. I want set the data pointer( malloced ) to CURL*, then when i before curl_easy_cleanup, i get the data pointer from CURL*, and free the data pointer, then call curl_easy_cleanup;

**Reply by xinsuinizhuan** [reply]
my data my be a structor or a class, include my send data and orther informatino about this send, and i new or malloc it, put them into queue. My thread to search the queque,  but when i remove, i only have the CURL*, no my data info, so i can't free it. So i hope CURL* have the usedata params, i can set, like callback function, then i could get it from CURL*.

**Reply by bagder** [reply]
Do you mean that you want to **retrieve** the pointer you previous set with `curl_easy_setopt()` ? I don't understand what "I want set the data pointer" refers to.

**Reply by xinsuinizhuan** [reply]
I define a struct HttpSendDataInfo:
~~~c
typedef struct _HttpSendDataInfo {
	CURL* curl;
	CURLM* mh;
	char* senddata_buf;
	struct curl_slist* pList;
	int senddata_sizel;
	BOOL prevent_cache;
	int submit_type;
	char urladdr[MAX_PATH];
	int time_out;
	BufData resultdata_info;
}HttpSendDataInfo;

HttpSendDataInfo* psenddatainfo = (HttpSendDataInfo*)malloc(sizeof(HttpSendDataInfo));
if (psenddatainfo == NULL) {
return 0;
}
~~~
i want use curl_easy_setopt() to set the data points(HttpSendDataInfo* psenddatainfo ) to CURL* params,
and last:
~~~c
int msgs_left = 0;
		CURLMsg* msg = NULL;
		while ((msg = curl_multi_info_read(g_async_multi, &msgs_left))) {
			if ((msg != NULL) && (msg->msg == CURLMSG_DONE)) {
				CURL* easy = msg->easy_handle;
				CURLcode res = msg->data.result;
				curl_multi_remove_handle(g_async_multi, easy);
				curl_easy_cleanup(easy);
				log_msg_ex("res=%d", res);
			}
		}
~~~
when i get the CURL* easy = msg->easy_handle, i could get the  data points(HttpSendDataInfo* psenddatainfo ), the i 
can free it.

**Comment by dfandrich** [comment]
You mean like 

curl_easy_setopt(easy, CURLOPT_PRIVATE, psenddatainfo);
…
curl_easy_getinfo(easy, CURLINFO_PRIVATE, &myptr);

?

**Reply by xinsuinizhuan** [reply]
Yes. when i use the curl_multi_info_read(g_async_multi, &msgs_left), get the CURL* easy = msg->easy_handle, i also get my data points, then i could free them.

**Reply by xinsuinizhuan** [reply]
Or add a param to CURL，could  assign the void*(data points) to this param, then when i get the CURL, i also get the data points, then to release

**Reply by bagder** [reply]
... like the options @dfandrich suggested do...

**Reply by xinsuinizhuan** [reply]
how about it？could support?

**Reply by xinsuinizhuan** [reply]
A add a  void* usedata, to  Curl_easy, but i could not visit it
![图片](https://user-images.githubusercontent.com/40679769/191880654-07836a82-f7d8-47f6-bf7c-7f9f631495bc.png)
void* usrdata = curl->usedata;

**Reply by xinsuinizhuan** [reply]
it also need:
curl_easy_setopt(easy, CURLOPT_PRIVATE, psenddatainfo);
…
curl_easy_getinfo(easy, CURLINFO_PRIVATE, &myptr);

**Comment by dfandrich** [comment]
I see some things that may be an issue. First, the type of the pointers passed
in do not match what the documentation says they must be. Also, there is no
transfer happening in the code you show, and curl_easy_getinfo says:

  Use this function AFTER a performed transfer if you want to get transfer
  related data.

I don't think CURLINFO_PRIVATE would be considered transfer-related data,
though.

**Reply by xinsuinizhuan** [reply]
I test it, now it seem ok:
~~~c
 BufData* recv_data = NULL;
  recv_data = (BufData*)malloc(sizeof(BufData));
  memset(recv_data, 0, sizeof(BufData));

  recv_data->buf = (char*)malloc(1920 * 1080 * 4);
  memset(recv_data->buf, 0, 1920 * 1080 * 4);

  recv_data->size = 1920 * 1080 * 4;
  recv_data->pos = 0;

  curl_easy_setopt(curl, CURLOPT_PRIVATE, recv_data);
  if(curl_multi_add_handle(mcurl, curl))
    goto fail;

  BufData* bufp = NULL;
  curl_easy_getinfo(curl, CURLINFO_PRIVATE, &bufp);

  if (bufp != NULL)
  {
      if (bufp->buf != NULL)
      {
		  free(bufp->buf);
          bufp->buf = NULL;
      }
	  free(bufp);
      bufp = NULL;
  }
